src/sync/impl/sync_client.hpp

```diff
@@ -41,21 +41,26 @@ using ReconnectMode = sync::Client::ReconnectMode;
 struct SyncClient {
     sync::Client client;
 
-    SyncClient(std::unique_ptr<util::Logger> logger,
-               ReconnectMode reconnect_mode = ReconnectMode::normal)
-    : client(make_client(*logger, reconnect_mode)) // Throws
-    , m_logger(std::move(logger))
-    , m_thread([this] {
-        if (g_binding_callback_thread_observer)
-            g_binding_callback_thread_observer->did_create_thread();
-
-        auto will_destroy_thread = util::make_scope_exit([&]() noexcept {
-            if (g_binding_callback_thread_observer)
-                g_binding_callback_thread_observer->will_destroy_thread();
-        });
-
-        client.run(); // Throws
-    }) // Throws
+    SyncClient(std::unique_ptr<util::Logger> logger, ReconnectMode reconnect_mode = ReconnectMode::normal)
+        : client(make_client(*logger, reconnect_mode)) // Throws
+        , m_logger(std::move(logger))
+        , m_thread([this] {
+            if (g_binding_callback_thread_observer) {
```

Comment:
This is getting to the point where it may be clearer to provide a default observer that does the appropriate actions so we don't have to worry about these conditionals. Maybe that's not worth it though since it's currently only used in this one place.